flask_resources/responses.py

```diff
@@ -9,77 +10,88 @@
+        return inner
 
-    def make_headers(self, content=None):
-        """Build response headers."""
-        return {"content-type": resource_requestctx.accept_mimetype}
+    return decorator
 
-    def make_item_response(self, content, code):
-        """Builds a response."""
-        raise NotImplementedError()
 
-    def make_list_response(self, content, code):
-        """Builds a response."""
-        raise NotImplementedError()
+class ResponseHandler:
+    """Response handler which delegates to the a serializer.
 
-    def make_error_response(self, error):
-        """Builds an error response."""
-        raise NotImplementedError()
+    Example usage:
 
+    .. code-block:: python
 
-class Response(ResponseMixin):
-    """Response implementation.
+        def obj_headers(obj_or_list, code, many=False):
+            return {'etag': ... }
 
-    Builds up a reponse for a single or list of objects.
+        class Config(ResourceConfig):
+            response_handlers = {
+                "application/json": ResponseHandler(
+                    JSONSerializer(), headers=obj_headers)
+            }
     """
 
-    def __init__(self, serializer=None):
+    def __init__(self, serializer, headers=None):
         """Constructor."""
         self.serializer = serializer
+        self.headers = headers
+
+    def make_headers(self, obj_or_list, code, many=False):
+        """Builds the headers fo the response."""
+        if self.headers is None:
+            return {
+                "content-type": resource_requestctx.accept_mimetype,
+            }
+        elif callable(self.headers):
+            return self.headers(obj_or_list, code, many=many)
+        else:
+            return self.headers
+
+    def make_response(self, obj_or_list, code, many=False):
+        """Builds a response for one object."""
+        # If view returns a response, bypass the serialization.
+        if isinstance(obj_or_list, Response):
+            return obj_or_list
 
-    def make_item_response(self, content, code=200):
-        """Builds a response for a single object."""
-        # https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.make_response
-        # (body, status, header)
-
-        return make_response(
-            "" if content is None else self.serializer.serialize_object(content),
-            code,
-            self.make_headers(content=content),
-        )
-
-    def make_list_response(self, content, code=200):
-        """Builds a response for a list of objects."""
         # https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.make_response
         # (body, status, header)
+        if many:
+            serialize = self.serializer.serialize_object_list
+        else:
+            serialize = self.serializer.serialize_object
 
         return make_response(
-            "" if content is None else self.serializer.serialize_object_list(content),
+            "" if obj_or_list is None else serialize(obj_or_list),
```

Comment:
Not sure we need this if anymore, given that you can selectively apply the response handlers.